### PR TITLE
Receive slice

### LINKED ISSUE #2: Receive Slice
# Headline
The wallet needs to be able to receive and hold a slice on behalf of a user, so the user has ownership and can later slice, transfer or claim his GCs.

# Implementation Checklist
- [ ] Design and document functionality
- [ ] Document Sequence diagram
- [ ] Write Protobuf scheme
- [ ] Implement code
- [ ] Write test

### PR DESCRIPTION
**Pull request by MartinSchmidt**
Implements the functionality to receive a slice into a wallet-section.

Features added:
- Support to receive a slice, fixes #2 
- Documentation of the receive-slice flow
- Added tables and migrations for Slices, Certificates and Registries

Not in scope:
- Verification of slices #5

### CODE REVIEW

File: doc/flows/receive_slice.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Slice receive flow
+
+Granular Certificates are "transferred" as slices to the wallet,
+the actual slice is on the registry,
+the necessary information to prove ownership and act on the slice is transferred.
+
+1. A source sends a slice to a recipient wallet.
+2. The wallet will try to look-up the wallet-section based on the public-key.
+3. If the certificate is not already know by the wallet, then an reference to it is inserted into the datastore
```
Comment (duizer) [inline doc/flows/receive_slice.md:9]:
```suggestion
3. If the certificate is not already known by the wallet, then a reference to the certificate is inserted into the datastore
```

File: doc/flows/receive_slice.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        alt Certificate not known
+            wallet ->> wallet: insert Certificate
+        end
+
+        wallet ->> wallet: insert Slice
+
+        wallet -->> source: Success
+    else wallet-section not found
+        wallet -->>- source: Success
```
Comment (duizer) [inline doc/flows/receive_slice.md:35]:
```suggestion
        wallet -->>- source: Error
```

File: doc/flows/receive_slice.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    link wallet: Dashboard @ https://dashboard.contoso.com/alice
+
+    loop Every minute
+        sche -->>+ wallet: VerifySlices
+
+        wallet ->> wallet: Get unverified slices
+
+        wallet ->> reg: Get Cert and Slice Info
+        reg ->> wallet: slice info
+        wallet ->> wallet: Verify Certificate inssuer
```
Comment (duizer) [inline doc/flows/receive_slice.md:66]:
```suggestion
        wallet ->> wallet: Verify Certificate issuer
```

File: doc/flows/receive_slice.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+A background process will then verify and update the state of the slice later. If the certificate and attributes are unknown, then they will be updated as part of this flow.
+
+```mermaid
+sequenceDiagram
+    autonumber
+
+    actor sche as Scheduler
+    participant wallet as Wallet
+    participant reg as Register
```
Comment (duizer) [inline doc/flows/receive_slice.md:55]:
```suggestion
    participant reg as Registry
```

File: doc/flows/receive_slice.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    link wallet: Dashboard @ https://dashboard.contoso.com/alice
```
Comment (duizer) [inline doc/flows/receive_slice.md:57]: What does this do?

Comment (MartinSchmidt) [reply doc/flows/receive_slice.md]: It was just something i tested, i have removed it.

File: src/ProjectOrigin.WalletSystem.Server/Database/Mapping/ApplicationBuilderExtension.cs [hunk 6]
```diff
@@ -11,7 +11,7 @@ public static void ConfigureSqlMappers(this IApplicationBuilder app)
     {
         var algorithm = app.ApplicationServices.GetRequiredService<IHDAlgorithm>();
 
-        SqlMapper.AddTypeHandler(new HDPrivateKeyTypeHandler(algorithm));
-        SqlMapper.AddTypeHandler(new HDPublicKeyTypeHandler(algorithm));
+        SqlMapper.AddTypeHandler<IHDPrivateKey>(new HDPrivateKeyTypeHandler(algorithm));
+        SqlMapper.AddTypeHandler<IHDPublicKey>(new HDPublicKeyTypeHandler(algorithm));
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.Server/Database/Mapping/ApplicationBuilderExtension.cs:15]: Generic arguments can be inferred. They are not needed.

Comment (MartinSchmidt) [reply src/ProjectOrigin.WalletSystem.Server/Database/Mapping/ApplicationBuilderExtension.cs]: done

File: src/ProjectOrigin.WalletSystem.IntegrationTests/Repositories/AbstractRepositoryTests.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    protected async Task<Registry> CreateRegistry()
+    {
+        var registry = new Registry(Guid.NewGuid(), Fixture.Create<string>());
+
+        using (var connection = CreateConnection())
+        {
+            await connection.ExecuteAsync("INSERT INTO Registries(Id, Name) VALUES (@id, @name)", registry);
+        }
+
+        return registry;
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.WalletSystem.IntegrationTests/Repositories/AbstractRepositoryTests.cs:55]: done

File: src/ProjectOrigin.WalletSystem.Server/Models/CertificateState.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+    /// <summary>
+    /// The certificate has been inserted, but header information has not yet been loaded from the registry.
+    /// </summary>
+    Inserted = 0,
+
+    /// <summary>
+    /// The certificate has been loaded from the registry and is valid.
+    /// </summary>
+    Valid = 100,
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.Server/Models/CertificateState.cs:19]: Is it needed to have ranges on the State? It is not exposed through the gRPC interface

File: src/ProjectOrigin.WalletSystem.Server/Models/Primitives/Period.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace ProjectOrigin.WalletSystem.Server.Models.Primitives
+{
+    public class Period
+    {
+        public DateTime DateFrom { get; }
+        public DateTime DateTo { get; }
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.Server/Models/Primitives/Period.cs:8]: Why not DateTimeOffset here?

File: src/ProjectOrigin.WalletSystem.Server/Models/SliceState.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    /// <summary>
+    /// The slice has not been verified yet.
+    /// </summary>
+    Unverified = 0,
+
+    /// <summary>
+    /// The slice has been verified and is valid.
+    /// </summary>
+    Valid = 100,
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.Server/Models/SliceState.cs:19]: Is it needed to have ranges on the State? It is not exposed through the gRPC interface

File: src/ProjectOrigin.WalletSystem.Server/Models/Primitives/Period.cs (hunk 9 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.WalletSystem.Server/Models/Primitives/Period.cs]: removed, was not used yet.

File: src/ProjectOrigin.WalletSystem.Server/Services/ExternalWalletService.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        var certId = Guid.Parse(federatedStreamId.StreamId.Value);
+        Registry? registry = await _unitOfWork.RegistryRepository.GetRegistryFromName(federatedStreamId.Registry);
+
+        if (registry == null)
+        {
+            registry = new Registry(Guid.NewGuid(), federatedStreamId.Registry);
+            await _unitOfWork.RegistryRepository.InsertRegistry(registry);
+        }
+
+        Certificate? certificate = await _unitOfWork.CertificateRepository.GetCertificate(registry.Id, certId);
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.Server/Services/ExternalWalletService.cs]:
```suggestion
        var certificate = await _unitOfWork.CertificateRepository.GetCertificate(registry.Id, certId);
```

File: src/ProjectOrigin.WalletSystem.IntegrationTests/ReceiveSliceTests.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            : base(grpcFixture, dbFixture, outputHelper)
+        {
+            _algorithm = new Secp256k1Algorithm();
+        }
+
+        private async Task<WalletSection> CreateWalletSection(string owner)
+        {
+            SqlMapper.AddTypeHandler<IHDPrivateKey>(new HDPrivateKeyTypeHandler(_algorithm));
+            SqlMapper.AddTypeHandler<IHDPublicKey>(new HDPublicKeyTypeHandler(_algorithm));
+
```
Comment (duizer) [inline src/ProjectOrigin.WalletSystem.IntegrationTests/ReceiveSliceTests.cs:33]: Is this needed?

**Pull request merged**